numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -2097,6 +2097,22 @@ array_fromfile(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *keywds)
         Py_DECREF(file);
         return NULL;
     }
+
+    PyObject *gzip_module = PyImport_ImportModule("gzip");
+    if(gzip_module == NULL) {
+        return NULL;
+    }
+    PyObject *gzip_class = PyObject_GetAttrString(gzip_module, "GzipFile");
+    Py_DECREF(gzip_module);
+    int is_gzip = PyObject_IsInstance(file, gzip_class);
+    Py_DECREF(gzip_class);
+    if (is_gzip == 1) {
+        PyErr_SetString(PyExc_TypeError, "'file' argument does not allow object of type GzipFile. Use frombuffer() instead.");
+        Py_XDECREF(type);
+        Py_DECREF(file);
+        return NULL;
+    }
```

Comment:
This should probably use `npy_cache_import`, the micro-optimizer in my, thinks that `PyObject_TypeCheck` may be better here.

But, overall I am curious if there is not some duck-typing way to find this? Although I guess `gzip` is common enough to be special? We do accept strings also, which will always be a direct file open... I suppose there is nothing we can do about that, though for very new users, I suspect that may be confusing.

Reply:
> But, overall I am curious if there is not some duck-typing way to find this?

Perhaps you could rely on the [magic number for gzip files](https://en.wikipedia.org/wiki/Gzip#File_format)? Such a check would likely be more light-weight, but I'm not sure what the tradeoff would be w.r.t. reliability.

Reply:
@seberg I have used `npy_cache_import` now although still using `PyObject_IsInstance` because I get a `PyObject *` from `npy_cache_import` and not a `PyType_Object *`

I think the current way to check type of Python file is better than using magic number. The issue with magic number is it might complain a binary file is gzip file if the binary file starts with `\x1f\x8b` and in the rare case this does happen, it will be almost impossible for the user to figure out why is the `fromfile` method complaining. I have added such a test also.